File: NeetCode/Swift/LinkedList/Medium/92.ReverseLinkedListII.py
# Given the head of a singly linked list and two integers left and 
# right where left <= right, reverse the nodes of the list from 
# position left to position right, and return the reversed list.

class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def reverseBetween(head, left, right):
    new_node = ListNode(val=0, next=head)
    previous, current = new_node, head

    for i in range(left - 1):
        previous = current 
        current = current.next

    last_node = None
    for i in range(right - left + 1):
        next_node = current.next
        current.next = last_node
        last_node = current
        current = next_node

    previous.next.next = current
    previous.next = last_node

    return new_node.next

File: swift/Neetcode/LinkedList/Medium/92.ReverseLinkedListII.swift
// Given the head of a singly linked list and two integers left and 
// right where left <= right, reverse the nodes of the list from 
// position left to position right, and return the reversed list.

func reverseBetween(_ head: ListNode?, _ left: Int, _ right: Int) -> ListNode? {
    let newHead = ListNode(0, head)
    var start: ListNode? = newHead

    for _ in 0..<(left - 1) {
        start = start?.next
    }

    var previous: ListNode? = nil, end = start?.next

    for _ in left...right {
        let next = end?.next
        end?.next = previous
        previous = end
        end = next
    }

    start?.next?.next = end
    start?.next = previous

    return newHead.next
}